Documentation/Heap.md

```diff
@@ -0,0 +1,148 @@
+
+There are a couple of options for initializing a `Heap`. To create an empty `Heap`, call `init()`:
+
+```swift
+var heap = Heap<Int>()
+```
+
+You can also create a `Heap` from an existing sequence in linear time:
+
+```swift
+var heap = Heap((1...).prefix(20))
+```
+
+Finally, a `Heap` can be created from an array literal:
+
+```swift
+var heap: Heap<Double> = [0.1, 0.6, 1.0, 0.15, 0.42]
+```
+
+### Insertion
+
+#### Of a single element
+
+To insert an element into a `Heap`, call `insert(_:)`:
+
+```swift
+var heap = Heap<Int>()
+heap.insert(6)
+heap.insert(2)
+```
+
+This works by adding the new element into the end of the backing array and then bubbling it up to where it belongs in the heap.
+
+#### Of a sequence of elements
+
+You can also insert a sequence of elements into a `Heap`:
+
+```swift
+var heap = Heap((0..<10))
+heap.insert(contentsOf: (20...100).shuffled())
+heap.insert(contentsOf: [-5, -6, -8, -12, -3])
+```
+
+### Lookup
+
+As mentioned earlier, the smallest and largest elements can be queried in constant time:
+
+```swift
+var heap = Heap((1...20))
+let min = heap.min()  // min = 1
+let max = heap.max()  // max = 20
+```
+
+In a min-max heap, the smallest element is stored at index 0 in the backing array; the largest element is stored at either index 1 or index 2, the first max level in the heap (so to look up the largest, we compare the two and return the larger one).
+
+We also expose a read-only view into the backing array, should somebody need that.
+
+```swift
+let heap = Heap((1...100).shuffled())
+for val in heap.unordered {
+   ...
+}
+```
+
+> Note: The elements aren't _arbitrarily_ ordered (it is, after all, a heap). However, no guarantees are given as to the ordering of the elements or that this won't change in future versions of the library.
+
+### Removal
+
+Removal has logarithmic complexity, and removing both the smallest and largest elements is supported:
+
+```swift
+var heap = Heap((1...20).shuffled())
+var heap2 = heap
+
+while let min = heap.popMin() {
+    print("Next smallest element:", min)
+}
+
+while let max = heap2.popMax() {
+    print("Next largest element:", max)
+}
+```
+
+To remove the smallest element, we remove and return the element at index 0. The element at the end of the backing array is put in its place at index 0, and then we trickle it down to where it belongs in the heap. To remove the largest element, we do the same except the index is whatever the index of the largest element is (see above) instead of 0.
+
+We also have non-optional flavors that assume the queue isn't empty, `removeMin()` and `removeMax()`.
+
+### Iteration
+
+`Heap` itself doesn't conform to `Sequence` because of the potential confusion around which direction it should iterate (largest-to-smallest? smallest-to-largest?). Instead, we expose two iterators that conform to `Sequence`:
+
+```swift
+for val in heap.ascending {
+    ...
+}
+
+for val in heap.descending {
+    ...
+}
+```
+
+### Performance
```

Comment:
Was there supposed to be any performance data under this header?

Reply:
Ooops. I added that because it was in most of the other docs. I didn't end up filling it in, as it's covered elsewhere in here. Thoughts?

Reply:
Perhaps just move the time complexity characteristics from the overview section there and basically list the operations with their time complexity characteristics? 

There is no precedent for adding performance graphs, but in general I think that could be useful if combined with hardware information - when looking at data structure repos, it is one of the first obvious questions for me at least, just to get an indication of approximate level of performance (as using terms as "fast" have very different meanings to different people in my experience - and algorithmic complexity doesn't say anything about constant overhead which might be troublesome depending on use case). So I'd consider attaching the perf data for just this data structure.

Reply:
I added a table with the performance of the various operations. I'll defer to @kylemacomber and @lorentey on the point about performance graphs. I agree that they can be helpful, but I also wonder what the baseline of comparison one would want to see in them — Comparing to `Foundation`/`Core Foundation` types? Comparing to C++'s stdlib types? Just average per-element operation time (i.e. only show the graphs for `Heap` itself)?

Reply:
Thanks! Speaking from my experience I’d be happy with just the graphs for Heap itself. It gives some orientation and if I really care about performance at that level, I’d already have some notion of how other similar structures perform.  I think the most important thing is to give the magnitude of performance numbers to give context - as we all have very widely different views of “fast” - this anchors that notion.

Reply:
Added